rados/read_op_omap_get_vals_by_keys.go

```diff
@@ -0,0 +1,124 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <rados/librados.h>
+// #include <stdlib.h>
+//
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/internal/cutil"
+)
+
+// ReadOpOmapGetValsByKeysStep holds the result of the
+// GetOmapValuesByKeys read operation.
+// Result is valid only after Operate() was called.
+type ReadOpOmapGetValsByKeysStep struct {
+	// C arguments
+
+	iter  C.rados_omap_iter_t
+	prval *C.int
+
+	// Internal state
+
+	// canIterate is only set after the operation is performed and is
+	// intended to prevent premature fetching of data.
+	canIterate bool
+}
+
+func newReadOpOmapGetValsByKeysStep(keys []string) *ReadOpOmapGetValsByKeysStep {
+	s := &ReadOpOmapGetValsByKeysStep{
+		prval: (*C.int)(C.malloc(C.sizeof_int)),
+	}
+
+	return s
+}
+
+func (s *ReadOpOmapGetValsByKeysStep) free() {
+	s.canIterate = false
+	C.rados_omap_get_end(s.iter)
+
+	C.free(unsafe.Pointer(s.prval))
+	s.prval = nil
+}
+
+func (s *ReadOpOmapGetValsByKeysStep) update() error {
+	err := getError(*s.prval)
+	s.canIterate = (err == nil)
+
+	return err
+}
+
+// Next gets the next omap key/value pair referenced by
+// ReadOpOmapGetValsByKeysStep's internal iterator.
+// If there are no more elements to retrieve, (nil, nil) is returned.
+// May be called only after Operate() finished.
+//  PREVIEW
+func (s *ReadOpOmapGetValsByKeysStep) Next() (*OmapKeyValue, error) {
+	if !s.canIterate {
+		return nil, ErrOperationIncomplete
+	}
+
+	var (
+		cKey    *C.char
+		cVal    *C.char
+		cValLen C.size_t
+	)
+
+	ret := C.rados_omap_get_next(s.iter, &cKey, &cVal, &cValLen)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	if cKey == nil {
+		// Iterator has reached the end of the list.
+		return nil, nil
+	}
+
+	return &OmapKeyValue{
+		Key:   C.GoString(cKey),
+		Value: C.GoBytes(unsafe.Pointer(cVal), C.int(cValLen)),
+	}, nil
+}
+
+// GetOmapValuesByKeys starts iterating over specific key/value pairs.
+//  PREVIEW
+//
+// Implements:
+//  void rados_read_op_omap_get_vals_by_keys(rados_read_op_t read_op,
+//                                           char const * const * keys,
+//                                           size_t keys_len,
+//                                           rados_omap_iter_t * iter,
+//                                           int * prval)
+func (r *ReadOp) GetOmapValuesByKeys(keys []string) *ReadOpOmapGetValsByKeysStep {
+	s := newReadOpOmapGetValsByKeysStep(keys)
+	r.steps = append(r.steps, s)
+
+	cKeys := cutil.NewCPtrCSlice(len(keys))
+	defer func() {
+		for _, cKeyPtr := range cKeys {
+			C.free(unsafe.Pointer(cKeyPtr))
+		}
+		cKeys.Free()
+	}()
+
+	for i, key := range keys {
+		cKey := C.CString(key)
+		cKeys[i] = cutil.CPtr(cKey)
+	}
```

Comment:
@ansiwen I've moved the C key slice over to `GetOmapValuesByKeys`. Do you reckon there is a safe way to populate `cKeys` with ptrs to individual keys without needing to do extra allocations with `C.CString`? I'm not sure there is, but better to ask...

Reply:
We _could_ do it, yes. With `unsafe.Pointer((*reflect.StringHeader)(unsafe.Pointer(&str)).Data)),` we could get the pointers to the start of the strings and put them in a pointer array, put the lengths in another array and then use bith with `rados_read_op_omap_get_vals_by_keys2()`. But then we also have to jump through other loops, because we have then Go pointers in a memory structure handed to C, that means we have to "pin" all these pointers. So, we would avoid the unnecessary copy, but the code wouldn't get any simpler, quite the opposite. Because we are not talking about big buffers here, but rather small key strings, I would suggest to simply copy the strings with `C.CString()` to keep the code simpler.

Reply:
Sounds reasonable, thanks!